GVFS/GVFS.Common/Maintenance/GitMaintenanceStep.cs

```diff
@@ -0,0 +1,172 @@
+        {
+            try
+            {
+                if (this.RequireObjectCacheLock)
+                {
+                    using (FileBasedLock cacheLock = GVFSPlatform.Instance.CreateFileBasedLock(
+                        this.Context.FileSystem,
+                        this.Context.Tracer,
+                        Path.Combine(this.Context.Enlistment.GitObjectsRoot, ObjectCacheLock)))
+                    {
+                        if (!cacheLock.TryAcquireLock())
+                        {
+                            this.Context.Tracer.RelatedInfo(this.Area + ": Skipping work since another process holds the lock");
+                            return;
+                        }
+
+                        this.CreateProcessAndRun();
+                    }
+                }
+                else
+                {
+                    this.CreateProcessAndRun();
+                }
+            }
+            catch (IOException e)
+            {
+                this.Context.Tracer.RelatedWarning(
+                    metadata: this.CreateEventMetadata(e),
+                    message: this.Area + ": IOException while running action: " + e.Message,
+                    keywords: Keywords.Telemetry);
+            }
+            catch (Exception e)
+            {
+                this.Context.Tracer.RelatedError(
+                    metadata: this.CreateEventMetadata(e),
+                    message: this.Area + ": Exception while running action: " + e.Message,
+                    keywords: Keywords.Telemetry);
+                Environment.Exit((int)ReturnCode.GenericError);
+            }
+        }
+
+        public void Stop()
+        {
+            lock (this.gitProcessLock)
+            {
+                this.Stopping = true;
+
+                GitProcess process = this.GitProcess;
+
+                if (process != null)
+                {
+                    if (process.TryKillRunningProcess())
+                    {
+                        this.Context.Tracer.RelatedEvent(
+                            EventLevel.Informational,
+                            this.Area + ": killed background Git process during " + nameof(this.Stop),
+                            metadata: null);
+                    }
+                    else
+                    {
+                        this.Context.Tracer.RelatedEvent(
+                            EventLevel.Informational,
+                            this.Area + ": failed to kill background Git process during " + nameof(this.Stop),
+                            metadata: null);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Implement this method perform the mainteance actions. If the object-cache lock is required
+        /// (as specified by <see cref="RequireObjectCacheLock"/>), then this step is not run unless we
+        /// hold the lock.
+        /// </summary>
+        protected abstract void RunGitAction();

```

Comment:
Should this be named PerformMaintenance? Currently it makes it sound like it will run one git action but in the post fetch it is running multiple.